Writing to hdf: "static assertion failed" with Eigen matrix, mapped from OpenCV

* HighFive => 2.1-dev (current master)
* Eigen => 3.3.7
* OpenCV => 4.1.2, 4.2.0-dev
* Operating System / Platform => Debian 10 x86_64 GNU/Linux (4.19.0-6-amd64)
* Compiler => g++ 8.3.0

##### Detailed description

I can do an `Eigen::Map` of a `cv::Mat`, to recover the data in the form of an `Eigen::Matrix`, but I get a 'static assertion failed' as soon as I use HighFive to store that data in an hdf file.

Since the mapping appears to work, I do not understand why `H5Easy::dump` behaves differently than with any another `Eigen::Matrix`.

##### Steps to reproduce
###### 1. Eigen::Map:
```cpp

double data[16] = {1,2,3,4, 5,6,7,8, 9,10,11,12, 13,14,15,16};
cv::Mat cvmat = cv::Mat_<double>(4,4, data);

// fixed dimensions, column-major
Eigen::Map<Eigen::Matrix4d> eigen_simple(reinterpret_cast<double*>(cvmat.data));
```

###### 2. Behaves as expected:
```cpp
std::cout << eigen_simple << '\n' << std::endl;
std::cout << eigen_simple.row(0) << '\n' << std::endl;
std::cout << eigen_simple.col(0) << '\n' << std::endl;
```
indeed prints
```
 1  5  9 13
 2  6 10 14
 3  7 11 15
 4  8 12 16

 1  5  9 13

1
2
3
4
```

###### 3. Won't compile once I try to write the data to an hdf file:
```cpp
H5Easy::File output_hdf("out_test.hdf", H5Easy::File::Overwrite);

// won't compile:
H5Easy::dump(output_hdf, "/Eigen_simple", eigen_simple, H5Easy::DumpMode::Overwrite);
```
Here's the error:
```
/usr/local/include/highfive/bits/H5Converter_misc.hpp: In instantiation of ‘HighFive::details::data_converter<Scalar, Enable>::data_converter(Scalar&, HighFive::DataSpace&) [with Scalar = Eigen::Map<Eigen::Matrix<double, 4, 4> >; Enable = void]’:
/usr/local/include/highfive/bits/H5Slice_traits_misc.hpp:227:44:   required from ‘void HighFive::SliceTraits< <template-parameter-1-1> >::write(const T&) [with T = Eigen::Map<Eigen::Matrix<double, 4, 4> >; Derivate = HighFive::DataSet]’
/usr/local/include/highfive/h5easy_bits/H5Easy_scalar.hpp:29:9:   required from ‘static HighFive::DataSet H5Easy::detail::scalar::dump_impl<T, E>::run(HighFive::File&, const string&, const T&) [with T = Eigen::Map<Eigen::Matrix<double, 4, 4> >; E = void; std::__cxx11::string = std::__cxx11::basic_string<char>]’
/usr/local/include/highfive/h5easy_bits/H5Easy_scalar.hpp:137:49:   required from ‘HighFive::DataSet H5Easy::dump(HighFive::File&, const string&, const T&, H5Easy::DumpMode) [with T = Eigen::Map<Eigen::Matrix<double, 4, 4> >; std::__cxx11::string = std::__cxx11::basic_string<char>]’
path/to/file/main.cpp:39:85:   required from here
/usr/local/include/highfive/bits/H5Converter_misc.hpp:121:52: error: static assertion failed: supported datatype should be an arithmetic value, a std::string or a container/array
         static_assert((std::is_arithmetic<Scalar>::value ||
                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                        std::is_enum<Scalar>::value ||
                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~
                        std::is_same<std::string, Scalar>::value),
                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

The message is the same as in #256, but note that I am *not using Boost* here (so #261 won't help); I actually opted out when compiling HighFive. Note also that the problem remains with the current master (which already includes patch #258).

##### Full example

```cpp
#include <iostream>
#include <Eigen/Eigen>
#include <opencv2/core/eigen.hpp>
#include <highfive/H5Easy.hpp>

namespace eigentools {
	template<class T>
	using MatrixX = Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>; 
}


int main()
{

	double data[16] = {1,2,3,4, 5,6,7,8, 9,10,11,12, 13,14,15,16};
	cv::Mat cvmat = cv::Mat_<double>(4,4, data);

	// fixed dimensions, column-major
	Eigen::Map<Eigen::Matrix4d> eigen_simple(reinterpret_cast<double*>(cvmat.data));
	std::cout << eigen_simple << '\n' << std::endl;

	// dynamic, row-major
	Eigen::Map<eigentools::MatrixX<double>> eigen_dynamic_rowmajor(cvmat.ptr<double>(), cvmat.rows, cvmat.cols); 
	std::cout << eigen_dynamic_rowmajor << '\n' << std::endl;

	// idem
	Eigen::Map<eigentools::MatrixX<double>> eigen_dynamic_rowmajor_alt(reinterpret_cast<double*>(cvmat.data), cvmat.rows, cvmat.cols);
	std::cout << eigen_dynamic_rowmajor_alt << '\n' << std::endl;

	H5Easy::File output_hdf("out_test.hdf", H5Easy::File::Overwrite);

	// So far, so good.

	// However, putting any of these will prevent the programme from compiling:

	H5Easy::dump(output_hdf, "/Eigen_simple", eigen_simple, H5Easy::DumpMode::Overwrite);
	H5Easy::dump(output_hdf, "/EigenRowMajorFromCV", eigen_dynamic_rowmajor, H5Easy::DumpMode::Overwrite);
	H5Easy::dump(output_hdf, "/EigenRowMajorFromCV_alt", eigen_dynamic_rowmajor_alt, H5Easy::DumpMode::Overwrite);

	return 0;
}
```

--- a/include/highfive/H5Easy.hpp
+++ b/include/highfive/H5Easy.hpp
@@ -110,6 +110,24 @@ inline DataSet dump(File& file,
 #endif
 
 ///
+/// \brief Write an Eigen::Map object to a new DataSet in an open HDF5 file.
+///
+/// \param file Writeable opened file
+/// \param path Path of the DataSet
+/// \param data eigen matrix to write
+/// \param mode Write mode
+///
+/// \return the newly created DataSet
+///
+#ifdef H5_USE_EIGEN
+template <class T>
+inline DataSet dump(File& file,
+                    const std::string& path,
+                    const Eigen::Map<T>& data,
+                    DumpMode mode = DumpMode::Create);
+#endif
+
+///
 /// \brief Write "xt::xarray<T>" to a new DataSet in an open HDF5 file.
 ///
 /// \param file A writeable opened HDF5 file

--- a/include/highfive/H5Easy.hpp
+++ b/include/highfive/H5Easy.hpp
@@ -92,6 +92,24 @@ inline DataSet dump(File& file,
 #endif
 
 ///
+/// \brief Write an Eigen::Ref object to a new DataSet in an open HDF5 file.
+///
+/// \param file Writeable opened file
+/// \param path Path of the DataSet
+/// \param data eigen matrix to write
+/// \param mode Write mode
+///
+/// \return the newly created DataSet
+///
+#ifdef H5_USE_EIGEN
+template <class T>
+inline DataSet dump(File& file,
+                    const std::string& path,
+                    const Eigen::Ref<T>& data,
+                    DumpMode mode = DumpMode::Create);
+#endif
+
+///
 /// \brief Write "xt::xarray<T>" to a new DataSet in an open HDF5 file.
 ///
 /// \param file A writeable opened HDF5 file

--- a/include/highfive/h5easy_bits/H5Easy_Eigen.hpp
+++ b/include/highfive/h5easy_bits/H5Easy_Eigen.hpp
@@ -22,14 +22,14 @@ namespace detail {
 namespace eigen {
 
 // return the shape of the "Eigen::Matrix" as size 1 or 2 "std::vector<size_t>"
-template <class C, int Rows, int Cols, int Options, int MaxRows, int MaxCols>
-inline std::vector<size_t> shape(const Eigen::Matrix<C,Rows,Cols,Options,MaxRows,MaxCols>& data)
+template <class C>
+inline std::vector<size_t> shape(const C& data)
 {
-    if (Rows == 1) {
+    if (std::decay_t<C>::RowsAtCompileTime == 1) {
         return {static_cast<size_t>(data.cols())};
     }
 
-    if (Cols == 1) {
+    if (std::decay_t<C>::ColsAtCompileTime == 1) {
         return {static_cast<size_t>(data.rows())};
     }
 
@@ -65,18 +65,17 @@ inline std::vector<Eigen::Index> shape(const File& file,
 
 // write to open DataSet of the correct size
 // (use Eigen::Ref to convert to RowMajor; no action if no conversion is needed)
-template <class C, int Rows, int Cols, int Options, int MaxRows, int MaxCols>
-inline void write(DataSet& dataset,
-                  const Eigen::Matrix<C,Rows,Cols,Options,MaxRows,MaxCols>& data)
+template <class C>
+inline void write(DataSet& dataset, const C& data)
 {
     Eigen::Ref<
         const Eigen::Matrix<
-            C,
-            Rows,
-            Cols,
-            Cols==1?Eigen::ColMajor:Eigen::RowMajor,
-            MaxRows,
-            MaxCols>,
+            typename std::decay_t<C>::Scalar,
+            std::decay_t<C>::RowsAtCompileTime,
+            std::decay_t<C>::ColsAtCompileTime,
+            std::decay_t<C>::ColsAtCompileTime==1?Eigen::ColMajor:Eigen::RowMajor,
+            std::decay_t<C>::MaxRowsAtCompileTime,
+            std::decay_t<C>::MaxColsAtCompileTime>,
         0,
         Eigen::InnerStride<1>> row_major(data);
 
@@ -84,40 +83,33 @@ inline void write(DataSet& dataset,
 }
 
 // create DataSet and write data
-template <class T>
-struct dump_impl
+template <class C>
+static DataSet dump_impl(File& file,
+                         const std::string& path,
+                         const C& data)
 {
-    template <class C, int Rows, int Cols, int Options, int MaxRows, int MaxCols>
-    static DataSet run(File& file,
-                       const std::string& path,
-                       const Eigen::Matrix<C,Rows,Cols,Options,MaxRows,MaxCols>& data)
-    {
-        detail::createGroupsToDataSet(file, path);
-        DataSet dataset = file.createDataSet<C>(path, DataSpace(shape(data)));
-        detail::eigen::write(dataset, data);
-        file.flush();
-        return dataset;
-    }
-};
+    using type = typename std::decay_t<C>::Scalar;
+    detail::createGroupsToDataSet(file, path);
+    DataSet dataset = file.createDataSet<type>(path, DataSpace(shape(data)));
+    detail::eigen::write(dataset, data);
+    file.flush();
+    return dataset;
+}
 
 // replace data of an existing DataSet of the correct size
-template <class T>
-struct overwrite_impl
+template <class C>
+static DataSet overwrite_impl(File& file,
+                              const std::string& path,
+                              const C& data)
 {
-    template <class C, int Rows, int Cols, int Options, int MaxRows, int MaxCols>
-    static DataSet run(File& file,
-                       const std::string& path,
-                       const Eigen::Matrix<C,Rows,Cols,Options,MaxRows,MaxCols>& data)
-    {
-        DataSet dataset = file.getDataSet(path);
-        if (dataset.getDimensions() != shape(data)) {
-            throw detail::error(file, path, "H5Easy::dump: Inconsistent dimensions");
-        }
-        detail::eigen::write(dataset, data);
-        file.flush();
-        return dataset;
+    DataSet dataset = file.getDataSet(path);
+    if (dataset.getDimensions() != shape(data)) {
+        throw detail::error(file, path, "H5Easy::dump: Inconsistent dimensions");
     }
-};
+    detail::eigen::write(dataset, data);
+    file.flush();
+    return dataset;
+}
 
 // load from DataSet
 // convert to ColMajor if needed (HDF5 always stores row-major)
@@ -170,11 +162,25 @@ inline DataSet dump(File& file,
                     DumpMode mode)
 {
     if (!file.exist(path)) {
-        return detail::eigen::dump_impl<
-            Eigen::Matrix<T,Rows,Cols,Options,MaxRows,MaxCols>>::run(file, path, data);
+        return detail::eigen::dump_impl(file, path, data);
     } else if (mode == DumpMode::Overwrite) {
-        return detail::eigen::overwrite_impl<
-            Eigen::Matrix<T,Rows,Cols,Options,MaxRows,MaxCols>>::run(file, path, data);
+        return detail::eigen::overwrite_impl(file, path, data);
+    } else {
+        throw detail::error(file, path, "H5Easy: path already exists");
+    }
+}
+
+// front-end
+template <class T>
+inline DataSet dump(File& file,
+                    const std::string& path,
+                    const Eigen::Ref<T>& data,
+                    DumpMode mode)
+{
+    if (!file.exist(path)) {
+        return detail::eigen::dump_impl(file, path, data);
+    } else if (mode == DumpMode::Overwrite) {
+        return detail::eigen::overwrite_impl(file, path, data);
     } else {
         throw detail::error(file, path, "H5Easy: path already exists");
     }

--- a/include/highfive/h5easy_bits/H5Easy_Eigen.hpp
+++ b/include/highfive/h5easy_bits/H5Easy_Eigen.hpp
@@ -137,6 +137,22 @@ struct load_impl
     }
 };
 
+// universal front-end (to minimise double code)
+template <class T>
+inline DataSet dump(File& file,
+                    const std::string& path,
+                    const T& data,
+                    DumpMode mode)
+{
+    if (!file.exist(path)) {
+        return detail::eigen::dump_impl(file, path, data);
+    } else if (mode == DumpMode::Overwrite) {
+        return detail::eigen::overwrite_impl(file, path, data);
+    } else {
+        throw detail::error(file, path, "H5Easy: path already exists");
+    }
+}
+
 }  // namespace eigen
 
 // front-end
@@ -161,13 +177,7 @@ inline DataSet dump(File& file,
                     const Eigen::Matrix<T,Rows,Cols,Options,MaxRows,MaxCols>& data,
                     DumpMode mode)
 {
-    if (!file.exist(path)) {
-        return detail::eigen::dump_impl(file, path, data);
-    } else if (mode == DumpMode::Overwrite) {
-        return detail::eigen::overwrite_impl(file, path, data);
-    } else {
-        throw detail::error(file, path, "H5Easy: path already exists");
-    }
+    return detail::eigen::dump(file, path, data, mode);
 }
 
 // front-end
@@ -177,13 +187,7 @@ inline DataSet dump(File& file,
                     const Eigen::Ref<T>& data,
                     DumpMode mode)
 {
-    if (!file.exist(path)) {
-        return detail::eigen::dump_impl(file, path, data);
-    } else if (mode == DumpMode::Overwrite) {
-        return detail::eigen::overwrite_impl(file, path, data);
-    } else {
-        throw detail::error(file, path, "H5Easy: path already exists");
-    }
+    return detail::eigen::dump(file, path, data, mode);
 }
 
 // front-end
@@ -193,13 +197,7 @@ inline DataSet dump(File& file,
                     const Eigen::Map<T>& data,
                     DumpMode mode)
 {
-    if (!file.exist(path)) {
-        return detail::eigen::dump_impl(file, path, data);
-    } else if (mode == DumpMode::Overwrite) {
-        return detail::eigen::overwrite_impl(file, path, data);
-    } else {
-        throw detail::error(file, path, "H5Easy: path already exists");
-    }
+    return detail::eigen::dump(file, path, data, mode);
 }
 
 }  // namespace H5Easy

--- a/include/highfive/h5easy_bits/H5Easy_Eigen.hpp
+++ b/include/highfive/h5easy_bits/H5Easy_Eigen.hpp
@@ -22,14 +22,14 @@ namespace detail {
 namespace eigen {
 
 // return the shape of the "Eigen::Matrix" as size 1 or 2 "std::vector<size_t>"
-template <class C>
-inline std::vector<size_t> shape(const C& data)
+template <class T>
+inline std::vector<size_t> shape(const T& data)
 {
-    if (std::decay_t<C>::RowsAtCompileTime == 1) {
+    if (std::decay_t<T>::RowsAtCompileTime == 1) {
         return {static_cast<size_t>(data.cols())};
     }
 
-    if (std::decay_t<C>::ColsAtCompileTime == 1) {
+    if (std::decay_t<T>::ColsAtCompileTime == 1) {
         return {static_cast<size_t>(data.rows())};
     }
 
@@ -65,17 +65,17 @@ inline std::vector<Eigen::Index> shape(const File& file,
 
 // write to open DataSet of the correct size
 // (use Eigen::Ref to convert to RowMajor; no action if no conversion is needed)
-template <class C>
-inline void write(DataSet& dataset, const C& data)
+template <class T>
+inline void write(DataSet& dataset, const T& data)
 {
     Eigen::Ref<
         const Eigen::Matrix<
-            typename std::decay_t<C>::Scalar,
-            std::decay_t<C>::RowsAtCompileTime,
-            std::decay_t<C>::ColsAtCompileTime,
-            std::decay_t<C>::ColsAtCompileTime==1?Eigen::ColMajor:Eigen::RowMajor,
-            std::decay_t<C>::MaxRowsAtCompileTime,
-            std::decay_t<C>::MaxColsAtCompileTime>,
+            typename std::decay_t<T>::Scalar,
+            std::decay_t<T>::RowsAtCompileTime,
+            std::decay_t<T>::ColsAtCompileTime,
+            std::decay_t<T>::ColsAtCompileTime==1?Eigen::ColMajor:Eigen::RowMajor,
+            std::decay_t<T>::MaxRowsAtCompileTime,
+            std::decay_t<T>::MaxColsAtCompileTime>,
         0,
         Eigen::InnerStride<1>> row_major(data);
 
@@ -83,12 +83,12 @@ inline void write(DataSet& dataset, const C& data)
 }
 
 // create DataSet and write data
-template <class C>
+template <class T>
 static DataSet dump_impl(File& file,
                          const std::string& path,
-                         const C& data)
+                         const T& data)
 {
-    using type = typename std::decay_t<C>::Scalar;
+    using type = typename std::decay_t<T>::Scalar;
     detail::createGroupsToDataSet(file, path);
     DataSet dataset = file.createDataSet<type>(path, DataSpace(shape(data)));
     detail::eigen::write(dataset, data);
@@ -97,10 +97,10 @@ static DataSet dump_impl(File& file,
 }
 
 // replace data of an existing DataSet of the correct size
-template <class C>
+template <class T>
 static DataSet overwrite_impl(File& file,
                               const std::string& path,
-                              const C& data)
+                              const T& data)
 {
     DataSet dataset = file.getDataSet(path);
     if (dataset.getDimensions() != shape(data)) {
@@ -175,6 +175,22 @@ template <class T>
 inline DataSet dump(File& file,
                     const std::string& path,
                     const Eigen::Ref<T>& data,
+                    DumpMode mode)
+{
+    if (!file.exist(path)) {
+        return detail::eigen::dump_impl(file, path, data);
+    } else if (mode == DumpMode::Overwrite) {
+        return detail::eigen::overwrite_impl(file, path, data);
+    } else {
+        throw detail::error(file, path, "H5Easy: path already exists");
+    }
+}
+
+// front-end
+template <class T>
+inline DataSet dump(File& file,
+                    const std::string& path,
+                    const Eigen::Map<T>& data,
                     DumpMode mode)
 {
     if (!file.exist(path)) {

--- a/tests/unit/tests_high_five_easy.cpp
+++ b/tests/unit/tests_high_five_easy.cpp
@@ -269,4 +269,18 @@ BOOST_AUTO_TEST_CASE(H5Easy_Eigen_VectorXRowMajor)
     BOOST_CHECK_EQUAL(A.isApprox(A_r), true);
     BOOST_CHECK_EQUAL(B.isApprox(B_r), true);
 }
+
+BOOST_AUTO_TEST_CASE(H5Easy_Eigen_Map)
+{
+    H5Easy::File file("test.h5", H5Easy::File::Overwrite);
+
+    std::vector<int> A = {1, 2, 3, 4, 5, 6, 7, 8, 9};
+    Eigen::Map<Eigen::VectorXi> mapped_vector(A.data(), static_cast<int>(A.size()));
+
+    H5Easy::dump(file, "/path/to/A", mapped_vector);
+
+    std::vector<int> A_r = H5Easy::load<std::vector<int>>(file, "/path/to/A");
+
+    BOOST_CHECK_EQUAL(A == A_r, true);
+}
 #endif

